feat(validation): Add catalog entry validation in controller

varshaprasad96 opened the pull request:
1. Validate export references in entry spec
2. Aggregate PermissionClaims and API resources info from
referenced APIExport to entry status

Signed-off-by: Vu Dinh <vudinh@outlook.com>

varshaprasad96 committed:
Hack: reconciler logic for verifying if APIExport exists

varshaprasad96 commented:
This is based on the CE api defined in commit (https://github.com/kcp-dev/catalog/pull/5/commits/6f259d69748e905020e03a5f5db92432fa1e4548) and is subject to change. Please hold from reviewing as this is not complete yet. 
/hold

ncdc commented:
If you don't mind, please pause working on this until we have solidified the API. Thanks!

dinhxuanvu committed:
Add CatalogEntry validation in controller

1. Validate export references in entry spec
2. Aggregate PermissionClaims and API resources info from
referenced APIExport to entry status

Signed-off-by: Vu Dinh <vudinh@outlook.com>

dinhxuanvu committed:
Add e2e test cases to test controller validation process

Signed-off-by: Vu Dinh <vudinh@outlook.com>

varshaprasad96 committed:
fix syntax errors

dinhxuanvu committed:
Fix e2e test cases

Signed-off-by: Vu Dinh <vudinh@outlook.com>

varshaprasad96 committed:
Fix Makefile changes

dinhxuanvu committed:
Fix Makefiles and few kcp resources files

Signed-off-by: Vu Dinh <vudinh@outlook.com>

dinhxuanvu committed:
Fix image-related issues on Makefile and deployment

Signed-off-by: Vu Dinh <vudinh@outlook.com>

varshaprasad96 committed:
Add permission model to apiexport

dinhxuanvu committed:
Fix e2e test cases implementation details

Signed-off-by: Vu Dinh <vudinh@outlook.com>

dinhxuanvu committed:
Add log level for debugging

Signed-off-by: Vu Dinh <vudinh@outlook.com>

varshaprasad96 committed:
Check if rbac is working

dinhxuanvu committed:
Fix resource crd issue

Signed-off-by: Vu Dinh <vudinh@outlook.com>

ncdc commented on main.go:
```diff
@@ -108,8 +140,59 @@ func main() {
+// APIExport's virtual workspace.
+func restConfigForAPIExport(ctx context.Context, cfg *rest.Config, apiExportName string) (*rest.Config, error) {
+	scheme := runtime.NewScheme()
+	// Add catalogEntry scheme
+	if err := catalogv1alpha1.AddToScheme(scheme); err != nil {
+		return nil, fmt.Errorf("error adding catalog.kcp.dev/v1alpha1 to scheme: %w", err)
+	}
```
Not needed?

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+
+func init() {
+	rand.Seed(time.Now().Unix())
+	flag.StringVar(&workspaceName, "workspace", "", "Workspace in which to run these tests.")
```
The test should create whatever workspace(s) it needs. And not use a flag for this.

ncdc replied:
@stevekuznetsov I wonder if we should try to extract some of the kcp e2e framework code into a separate repo at some point, so it could be reused

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+
+func loadClusterConfig(t *testing.T, clusterName logicalcluster.Name) *rest.Config {
+	t.Helper()
+	restConfig, err := config.GetConfigWithContext("base")
```
I'm not sure we want to require a kubeconfig with a context named `base` for these tests to pass? Could we assume the current context in the kubeconfig is sufficient?

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+	}
+	c := loadClient(t, parent)
+	t.Logf("creating workspace %q", clusterName)
+	if err := c.Create(context.TODO(), &tenancyv1alpha1.ClusterWorkspace{
```
Global comment: don't use `context.TODO()` unless there is absolutely no way to get access to a preexisting context. The pattern we've established in the kcp e2e tests is that each `TestXYZ()` function does this near the top:

```go
ctx, cancel := context.WithCancel(context.Background())
t.Cleanup(cancel)
`

If you need both `t` and `ctx` in a helper function, you can pass them both in.

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+}
+
+func createAPIBinding(t *testing.T, workspaceCluster logicalcluster.Name) client.Client {
+	c := loadClient(t, workspaceCluster)
```
I'd have both this function and createWorkspace take in a client as a parameter, and construct the client once in the `TestXYZ()` functions as needed.

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+		return fmt.Errorf("could not create APIResourceSchema %q|%q: %v", workspaceCluster, apiName, err)
+	}
+
+	t.Logf("waiting for APIResourceSchema %q|%q to be found", workspaceCluster, apiName)
+	var apiResourceSchema apisv1alpha1.APIResourceSchema
+	if err := wait.PollImmediate(100*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+		fetchErr := c.Get(context.TODO(), client.ObjectKey{Name: apiName}, &apiResourceSchema)
+		if fetchErr != nil {
+			t.Logf("failed to get APIResourceSchema %q|%q: %v", workspaceCluster, apiName, err)
+			return false, fetchErr
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("APIResourceSchema %q|%q not found: %v", workspaceCluster, apiName, err)
+	}
```
This isn't strictly required - we know that if we created a thing, it is immediately gettable

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+		return fmt.Errorf("could not create APIExport %q|%q: %v", workspaceCluster, apiName, err)
+	}
+
+	t.Logf("waiting for APIExport %q|%q to be found", workspaceCluster, apiName)
+	var apiExport apisv1alpha1.APIExport
+	if err := wait.PollImmediate(100*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+		fetchErr := c.Get(context.TODO(), client.ObjectKey{Name: apiName}, &apiExport)
+		if fetchErr != nil {
+			t.Logf("failed to get APIExport %q|%q: %v", workspaceCluster, apiName, err)
+			return false, fetchErr
+		}
+		return true, nil
+	}); err != nil {
+		return fmt.Errorf("APIExport %q|%q not found: %v", workspaceCluster, apiName, err)
+	}
```
Can remove

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+		return nil, fmt.Errorf("could not create CatalogEntry %q|%q: %v", workspaceCluster, entryName, err)
+	}
+
+	t.Logf("waiting for CatalogEntry %q|%q to be found", workspaceCluster, entryName)
```
```suggestion
	t.Logf("waiting for CatalogEntry %q|%q to have APIExportValid condition", workspaceCluster, entryName)
```

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+		}
+		return true, nil
+	}); err != nil {
+		return nil, fmt.Errorf("CatalogEntry %q|%q not found: %v", workspaceCluster, entryName, err)
```
```suggestion
		return nil, fmt.Errorf("error waiting for CatalogEntry %q|%q status: %v", workspaceCluster, entryName, err)
```

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+
+	t.Logf("waiting for CatalogEntry %q|%q to be found", workspaceCluster, entryName)
+	var catalogEntry catalogv1alpha1.CatalogEntry
+	if err := wait.PollImmediate(100*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
```
I would move this into the `Test()` functions themselves

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+		if err != nil {
+			t.Fatalf("unable to create CatalogEntry: %v", err)
+		}
+
```
Insert polling loop to get entry and expect `conditions.IsTrue(entry, catalogv1alpha1.APIExportValidType)`

ncdc commented on test/e2e/controller_test.go:
```diff
@@ -0,0 +1,401 @@
+		if err != nil {
+			t.Fatalf("catalogEntry %q failed to be reconciled in cluster %q: %v", entry.GetName(), workspaceCluster, err)
+		}
+
```
Insert polling loop

varshaprasad96 committed:
Fix path cannot be added error
